File: String-Operations/Safe-Text-Split/Python/2.py
import sys
import os
import time

output_dir = '/tmp'
chunk_size = 4096

def split(file_path, output_dir, max_file_size):
    with open(file_path, "rb") as file:
        files_written = 0
        current_file_size = 0

        file_name = os.path.basename(file_path)

        # Read the input file in chunks
        while True:
            chunk = file.read(chunk_size)

            if not chunk:
                # Reached the end of the file
                break

            current_file_size += len(chunk)

            if current_file_size > max_file_size:
                # Split the chunk if the maximum file size is exceeded
                last_newline_index = chunk.rfind(b'\n', 0, max_file_size)
                split_chunk = chunk[:last_newline_index + 1]

                # Write the split chunk to a new file
                output_file_path = f'{output_dir}/{file_name}_{files_written}.txt'
                with open(output_file_path, "wb") as output_file:
                    output_file.write(split_chunk)

                # Prepare for the next file
                chunk = chunk[last_newline_index + 1:]
                files_written += 1
                current_file_size = len(chunk)

            if current_file_size > 0:
                # Write the remaining chunk to the current file
                output_file_path = f'{output_dir}/{file_name}_{files_written}.txt'
                with open(output_file_path, "ab") as output_file:
                    output_file.write(chunk)

                current_file_size = len(chunk)

if len(sys.argv) < 3:
    print("Please provide the input file name or file path and maximum file size.")
    sys.exit(1)

file_path = sys.argv[1]
max_file_size = int(sys.argv[2])

start = time.time_ns()
split(file_path, output_dir, max_file_size)
duration = time.time_ns() - start
print(f"Execution time: {duration} ns")

File: String-Operations/Safe-Text-Split/Javascript/2.js
const fs = require('fs');
const path = require('path');

const output_dir = '/tmp';
const chunk_size = 4096;

function split(file_path, output_dir, max_file_size) {
  const file = fs.openSync(file_path, 'r');
  let files_written = 0;
  let current_file_size = 0;

  const file_name = path.basename(file_path);
  let chunk = Buffer.alloc(chunk_size);

  while (true) {
    const bytesRead = fs.readSync(file, chunk, 0, chunk_size, null);

    if (bytesRead === 0) {
      break;
    }

    current_file_size += bytesRead;

    if (current_file_size > max_file_size) {
      const last_newline_index = chunk.lastIndexOf('\n', max_file_size);
      const split_chunk = chunk.slice(0, last_newline_index + 1);

      const output_file_path = path.join(output_dir, `${file_name}_${files_written}.txt`);
      fs.writeFileSync(output_file_path, split_chunk);

      const remaining_chunk = chunk.slice(last_newline_index + 1);
      files_written += 1;
      current_file_size = remaining_chunk.length;

      if (current_file_size > 0) {
        fs.appendFileSync(output_file_path, remaining_chunk);
      }
    } else if (current_file_size > 0) {
      const output_file_path = path.join(output_dir, `${file_name}_${files_written}.txt`);
      fs.appendFileSync(output_file_path, chunk);
    }
  }

  fs.closeSync(file);
}

if (process.argv.length < 4) {
  console.log('Please provide the input file name or file path and maximum file size.');
  process.exit(1);
}

const file_path = process.argv[2];
const max_file_size = parseInt(process.argv[3], 10);

const start = process.hrtime.bigint();
split(file_path, output_dir, max_file_size);
const end = process.hrtime.bigint();
const duration = end - start;
console.log(`Execution time: ${duration} ns`);
